### PR TITLE
feat: allow injecting a cache implementation that is used by the mapper

### PR DESCRIPTION
**Pull request by romm**
The cache implementation that was previously injected in the mapper
builder must now be manually injected. This gives better control on when
the cache should be enabled, especially depending on which environment
the application is running.

The library provides a cache implementation out of the box, which saves
cache entries in the file system.

It is also possible to use any PSR-16 compliant implementation, as long
as it is capable of caching the entries handled by the library.

When the application runs in development environment, the cache
implementation should be decorated with `FileWatchingCache` to prevent
invalid cache entries states, which can result in the library not
behaving as expected (missing property value, callable with outdated
signature, …).

```php
$cache = new \CuyZ\Valinor\Cache\FileSystemCache('path/to/cache-dir');

if ($isApplicationInDevelopmentEnvironment) {
    $cache = new \CuyZ\Valinor\Cache\FileWatchingCache($cache);
}

(new \CuyZ\Valinor\MapperBuilder())
    ->withCache($cache)
    ->mapper()
    ->map(SomeClass::class, [/* … */]);
```

This behavior now forces to explicitly inject `FileWatchingCache`, when
it was done automatically before; but because it shouldn't be used in
a production environment, it will increase overall performance.

### CODE REVIEW

File: src/Cache/FileWatchingCache.php [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+ * It should decorate the original cache implementation and should be given to
+ * the mapper builder: @see \CuyZ\Valinor\MapperBuilder::withCache
+ *
+ * @api
+ *
+ * @phpstan-type TimestampsArray = array<string, int>
+ * @template EntryType
+ * @implements CacheInterface<EntryType|TimestampsArray>
+ */
+final class FileWatchingCache implements CacheInterface
```
[the remark this answers is not in the document]
Comment (romm) [reply src/Cache/FileWatchingCache.php:33]: `Source` is already a keyword in the library, maybe `ApplicationFileWatchingCache`? WDYT?

[the remark this answers is not in the document]
Comment (Mopolo) [reply src/Cache/FileWatchingCache.php:33]: Oh right, then let's go for `ApplicationFileWatchingCache`.

[the remark this answers is not in the document]
Comment (Mopolo) [reply src/Cache/FileWatchingCache.php:33]: Yeah I agree haha. The doc change seems like a better idea.

**Comment by romm** [comment]
Looks good to you?

**Review by Mopolo — APPROVED** [review]

**Pull request merged**